src/unix/Lwt_fmt.mli

```diff
@@ -0,0 +1,98 @@
+(* OCaml promise library
+ * http://www.ocsigen.org/lwt
+ * Copyright (C) 2018 Gabriel Radanne
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, with linking exceptions;
+ * either version 2.1 of the License, or (at your option) any later
+ * version. See COPYING file for details.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ * 02111-1307, USA.
+ *)
+
+(** Format API for Lwt-powered IOs *)
+
+(** This module bridges the gap between [Format] and [Lwt].
+    Although it is not required, it is recommended to use this module with the
+    [Fmt] library.
+
+    Compared to regular formatting function, the main difference is that
+    printing statements will now return promises instead of blocking.
+*)
+
+val printf : ('a, Format.formatter, unit, unit Lwt.t) format4 -> 'a
+(** Returns a promise that prints on the standard output. 
+    Similar to [Format.printf]. *)
+
+val eprintf : ('a, Format.formatter, unit, unit Lwt.t) format4 -> 'a
+(** Returns a promise that prints on the standard error. 
+    Similar to [Format.eprintf]. *)
+
+(** {1 Formatters} *)
+
+type formatter
+(** Lwt enabled formatters *)
+
+type order =
+  | String of string * int * int (** [String (s, off, len)] indicate the output of [s] at offset [off] and length [len]. *)
+  | Flush (** Flush operation *)
+
+val make_stream : unit -> order Lwt_stream.t * formatter
+(** [make_stream ()] returns a formatter and a stream of all the writing
+    order given on that stream.
+*)
+
+
+val of_channel : Lwt_io.output_channel -> formatter
+(** [of_channel oc] creates a formatter that writes to the channel [oc]. *)
+
+val stdout : formatter (** Formatter printing on {!Lwt_io.stdout}. *)
+val stderr : formatter (** Formatter printing on {!Lwt_io.stdout}. *)
+
+val make_formatter :
+  commit:(unit -> unit Lwt.t) -> fmt:Format.formatter -> formatter
+(** [make_formatter ~commit ~fmt] creates a new lwt formatter based on the
+    {!Format.formatter} [fmt]. The [commit] function will be called by the printing
+    functions to update the underlying channel.
+*)
+
+(** {2 Printing} *)
+
+val fprintf : formatter -> ('a, Format.formatter, unit, unit Lwt.t) format4 -> 'a
+
+val kfprintf :
+  (Format.formatter -> unit Lwt.t -> 'a) ->
```

Comment:
I think it would make sense to have `(formatter -> 'a)` rather than `(Format.formatter -> unit Lwt.t -> 'a)` for the continuation argument:  with `write_pending` exposed, it is still possible to write `fprintf` from the `formatter -> 'a` version and it avoids exposing the underlying `Format.formatter` of the `formatter` argument.

Reply:
Hum, that is right, but I'm pretty sure people will forget to call `write_pending` if you don't do it for them, which will lead to very weird bugs.

I'm trying to think if exposing the `Format.formatter` is a bad thing or not. It's not *really* unsafe, but I'm not sure if it's useful or not.

Reply:
In term of coherency of the API, I think that in any case it would be better to expose a `formatter` rather than `Format.formatter` otherwise `kfprintf`  provides a strange entry point to features only available for `Format.formatter` (e.g. modifying margin, tags).

I agree that people will probably forget the call to `write_pending`, even if this could be mitigated by adding `Do not forget to call {!write_pending}` to the documentation and the limited amount of people using kprintf without reading the documentation.

Reply:
Or maybe we can keep the `unit Lwt.t` and just expose a `formatter`.

Do you think I need to expose a function to retrieve the `Format.formatter` too ?

Reply:
Yep, I think that `formatter -> unit Lwt.t -> `a` is a good compromise.

> Do you think I need to expose a function to retrieve the Format.formatter too ?

I think that It would be nice to avoid it. I believe that letting `make_formatter` handle the configuration of the formatter is fine since *safely* modifying the underlying formatter is as hard as *safely* creating a second formatter with the same output channel. Nevertheless, in this case, it might be interesting to add the part of the `Format` API not exposed to `kfprintf`, i.e. the tabulation boxes and `pp_print_if_newline`.

Reply:
Isn't that point sort of moot, given the existence of `%t` and `%a`? `Lwt_fmt.fprintf fmt "%t" (fun fmt -> ....)` will always work.

Reply:
True,`%t` and `%a` should be enough for tabulation boxes and friends.